umbraco/Umbraco-CMS
# V8: Fix tree flickering when switching between users with different start nodes

### Prerequisites


If there's an existing issue for this PR then this fixes https://github.com/umbraco/Umbraco-CMS/issues/6043

### Description

#6043 describes how the content tree momentarily displays the nodes visible to the previous user when a new user logs in. Steps to reproduce are given in the original issue description.

Long story short: It's not a caching issue. The built-in cache layers in Umbraco are flushed correctly when a new user logs in, and this causes the tree to reload correctly. 

The problem is that the previously loaded tree is still in the DOM when the new user logs in; the `umbTree` directive keeps the last loaded tree nodes in memory, and these remain rendered until the cache is cleared and the tree is re-rendered.

I can't think of any other solution to this issue than to explicitly unload the previously loaded tree upon logout, which is what this PR is all about.

There is no significant performance downside to this; the cache remains intact if the same user logs in and out (or times out), thus the tree isn't reloaded from the server and the previously known tree state (open nodes etc) isn't lost.

Here's how things work with this PR applied:

![login-flicker](https://user-images.githubusercontent.com/7405322/62679788-cd886d00-b9b5-11e9-94e3-be3127c12c78.gif)

#### To do or not to do?

This PR certainly introduces a bit of extra complexity to fix an issue that is likely to be a rare occurrence. One could also argue that the issue isn't that grave, since the tree *is* reloaded correctly after log-in - although it may not be as quick to reload in a real life scenario.

#### Testing this PR

1. Make sure the flickering is gone when switching between different users in the same browser window.
2. Make sure the tree state remains the same between logins with the same user.
3. Make sure that the user is redirected back to the content section when an explicit logout is performed from any other section.
4. Make sure both the tree state and the current section remains the same when the user times out (hint: set the `Umbraco.Core.TimeOutInMinutes` app setting value to 1 for testing).
